consul-lambda-registrator/extension/extension.go

```diff
@@ -0,0 +1,312 @@
+		case err := <-pErrChan:
+			if err != nil {
+				errChan <- fmt.Errorf("proxy failed with an error: %w", err)
+				return
+			}
+		}
+	}
+}
+
+func (ext *Extension) runEvents(ctx context.Context, errChan chan error) {
+	ext.Logger.Info("processing events")
+	err := ext.Events.ProcessEvents(ctx)
+	if err != nil {
+		errChan <- fmt.Errorf("event processing failed with an error: %w", err)
+		return
+	}
+	ext.Logger.Info("event processing completed")
+	errChan <- nil
+}
+
+func (ext *Extension) getExtensionData(ctx context.Context) (structs.ExtensionData, error) {
+	trace.Enter()
+	defer trace.Exit()
+
+	var extData structs.ExtensionData
+
+	// Retrieve the data.
+	key := fmt.Sprintf("%s%s", ext.ExtensionDataPrefix, ext.service.ExtensionPath())
+	d, err := ext.Store.Get(ctx, key)
+	if err != nil {
+		return extData, fmt.Errorf("failed to get extension data for %s: %w", key, err)
+	}
+
+	// Unmarshal.
+	if err := json.Unmarshal([]byte(d), &extData); err != nil {
+		return extData, fmt.Errorf("failed to unmarshal extension data for %s: %w", key, err)
+	}
+	return extData, nil
+}
+
+// startProxy starts, or restarts, the extension's proxy server.
+// It retrieves the configuration for the proxy and if the configuration has changed
+// it closes the existing proxy server and reconfigures a new proxy server.
+// It starts the new proxy server in a separate go routine that reports any errors to the
+// caller via errChan.
+func (ext *Extension) startProxy(ctx context.Context, errChan chan error) error {
+	trace.Enter()
+	defer trace.Exit()
+
+	const errFmt = "failed to init proxy: %w"
+
+	// Get the lambda extension configuration data for this service.
+	extData, err := ext.getExtensionData(ctx)
+	if err != nil {
+		// TODO: We need to handle the case where the data used to exist but it was deleted because
+		// this Lambda function was removed from the service mesh. In that case we should shut down
+		// the proxy server so the function can't make any more outgoing calls.
+		return fmt.Errorf(errFmt, err)
+	}
+
+	// If the extension data has not changed then the proxy is already configured.
+	if extData.Equals(ext.data) {
+		return nil
+	}
+
+	ext.Logger.Info("starting proxy server")
+
+	// Parse the configured list of upstreams. The upstreams are configured as part of the environment
+	// so we only do this on the first time through.
+	if len(ext.upstreams) == 0 {
+		ext.upstreams, err = ext.parseUpstreams()
+		if err != nil {
+			return fmt.Errorf(errFmt, err)
+		}
+	}
+
+	// Create a proxy listener configuration for each upstream.
+	proxyConfigs := make([]*proxy.Config, len(ext.upstreams))
+	for i, upstream := range ext.upstreams {
+		// Create the listener config.
+		cfg, err := ext.proxyConfig(upstream, extData)
+		if err != nil {
+			return fmt.Errorf(errFmt, err)
+		}
+		proxyConfigs[i] = cfg
+	}
+
+	// Drain and close the existing proxy.
+	ext.closeProxy()
+
+	// Create the proxy server.
+	ext.proxy = proxy.New(ext.Logger, proxyConfigs...)
```

Comment:
Is it possible to do this without shutting down the old proxies? Maybe by updating the proxy config data it references?

Reply:
Yes it's possible to reconfig without shutting down the proxies but then you have to manage draining any open connections. By using the proxy's `Close` method it automatically blocks until any open connections are closed without force closing them. I'm open to exploring doing the reconfig in-flight but I did it this way for the sake of time. Are you OK if I defer this until after the beta release?

Reply:
I've added an issue to investigate this post-beta.

Reply:
Yeah, post-beta is just fine.